plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts

```diff
@@ -0,0 +1,63 @@
+/**
+ * External dependencies
+ */
+import { dispatch, select } from '@wordpress/data';
+import { ShippingAddress } from '@woocommerce/settings';
+import { VALIDATION_STORE_KEY } from '@woocommerce/block-data';
+import { __ } from '@wordpress/i18n';
+
+function previousString( initialValue?: string ) {
+	let lastValue = initialValue;
+
+	function track( value: string ) {
+		const currentValue = lastValue;
+
+		lastValue = value;
+
+		// Return the previous value
+		return currentValue;
+	}
+
+	return track;
+}
+
+const lastValue = previousString( '' );
+
+export const validateState = (
+	addressType: string,
+	values: ShippingAddress,
+	isRequired: boolean
+) => {
+	const validationErrorId = `${ addressType }_state`;
+	const hasValidationError =
+		select( VALIDATION_STORE_KEY ).getValidationError( validationErrorId );
+
+	const countryChanged = lastValue( values.country ) !== values.country;
+
+	if ( hasValidationError ) {
+		if ( ! isRequired || values.state ) {
+			// Validation error has been set, but its no longer required, or the state was provided, clear the error.
+			dispatch( VALIDATION_STORE_KEY ).clearValidationError(
+				validationErrorId
+			);
+		} else if ( ! countryChanged ) {
```

Comment:
The comment should say "there has not been a country change" that distinguishing factor is important here. I will fix the comment, but I couldn't find a way to make this work without determining if the country has changed since the function was last run.

Reply:
@opr I've adjusted this logic because I found one case where it didn't work as expected when you change the address and it shouldn't show validation for the state yet. I've adjusted the comment to better match as well

[adjust validation to handle any address change, not just country. tra…](https://github.com/woocommerce/woocommerce/pull/49616/commits/662657168ad158e0ecde07e07dfd14de6f30214c) 
6626571